src/audio/module_adapter/module/volume/volume_common.c

```diff
@@ -0,0 +1,326 @@
+	cd->is_passthrough = false;
+}
+
+/**
+ * \brief Sets channel target volume.
+ * \param[in,out] mod Volume processing module handle
+ * \param[in] chan Channel number.
+ * \param[in] vol Target volume.
+ * \param[in] constant_rate_ramp When true do a constant rate
+ *	      and variable time length ramp. When false do
+ *	      a fixed length and variable rate ramp.
+ */
+int volume_set_chan(struct processing_module *mod, int chan,
+		    int32_t vol, bool constant_rate_ramp)
+{
+	struct vol_data *cd = module_get_private_data(mod);
+	int32_t v = vol;
+	int32_t delta;
+	int32_t delta_abs;
+	int32_t coef;
+
+	/* Limit received volume gain to MIN..MAX range before applying it.
+	 * MAX is needed for now for the generic C gain arithmetic to prevent
+	 * multiplication overflow with the 32 bit value. Non-zero MIN option
+	 * can be useful to prevent totally muted small volume gain.
+	 */
+	if (v < cd->vol_min) {
+		/* No need to fail, just trace the event. */
+		comp_warn(mod->dev, "volume_set_chan: Limited request %d to min. %d",
+			  v, cd->vol_min);
+		v = cd->vol_min;
+	}
+
+	if (v > cd->vol_max) {
+		/* No need to fail, just trace the event. */
+		comp_warn(mod->dev, "volume_set_chan: Limited request %d to max. %d",
+			  v, cd->vol_max);
+		v = cd->vol_max;
+	}
+
+	cd->tvolume[chan] = v;
+	cd->rvolume[chan] = cd->volume[chan];
+	cd->vol_ramp_elapsed_frames = 0;
+
+	/* Check ramp type */
+	if (cd->ramp_type == SOF_VOLUME_LINEAR ||
+	    cd->ramp_type == SOF_VOLUME_LINEAR_ZC) {
+		/* Get volume transition delta and absolute value */
+		delta = cd->tvolume[chan] - cd->volume[chan];
+		delta_abs = ABS(delta);
+
+		/* The ramp length (initial_ramp [ms]) describes time of mute
+		 * to vol_max unmuting. Normally the volume ramp has a
+		 * constant linear slope defined this way and variable
+		 * completion time. However in streaming start it is feasible
+		 * to apply the entire topology defined ramp time to unmute to
+		 * any used volume. In this case the ramp rate is not constant.
+		 * Note also the legacy mode without known vol_ramp_range where
+		 * the volume transition always uses the topology defined time.
+		 */
+		if (cd->initial_ramp > 0) {
+			if (constant_rate_ramp && cd->vol_ramp_range > 0)
+				coef = cd->vol_ramp_range;
+			else
+				coef = delta_abs;
+
+			/* Divide and round to nearest. Note that there will
+			 * be some accumulated error in ramp time the longer
+			 * the ramp and the smaller the transition is.
+			 */
+			coef = (2 * coef / cd->initial_ramp + 1) >> 1;
+		} else {
+			coef = delta_abs;
+		}
+
+		/* Scale coefficient by 1/8, round */
+		coef = ((coef >> 2) + 1) >> 1;
+
+		/* Ensure ramp coefficient is at least min. non-zero
+		 * fractional value.
+		 */
+		coef = MAX(coef, 1);
+
+		/* Invert sign for volume down ramp step */
+		if (delta < 0)
+			coef = -coef;
+
+		cd->ramp_coef[chan] = coef;
+		comp_dbg(mod->dev, "cd->ramp_coef[%d] = %d", chan, cd->ramp_coef[chan]);
+	}
+
+	return 0;
+}
```

Comment:
it should be larger, about 3 times as large. Please see attached example. Note, it isn't complete - I just replaced some IPC[34]-specific code with dummy function calls, you'd need to implement those properly and add respective parameters
[volume.c.zip](https://github.com/thesofproject/sof/files/12415809/volume.c.zip)

Reply:
do you mean move more code to common source file?  from my opinion, common file just used to save exactly same code, not all possible common code with more modification on both side, as long as ipc3 and ipc4 are totally separated, and ipc3 ipc4 source file is in right size(<1000), it should be ok?

Reply:
Hmm, looking at @lyakh 's code, so I understand right the proposal that we'd define function pointers for things like "vol_get_processing_function()" (with some optional parameters in the function template). So then we could have common code for e.g. prepare() (like in Guennadi's example), and only have the ops setup in ipc3/ipc4.c. I think this sounds reasonable, but devil may be in the details.

Reply:
Let's me recall the initial purpose of this PR:
volume.c is too large with around 1700 lines code mixed with IPC3 and IPC4

So, this PR main purpose is: 
1. separate IPC3 and IPC4 for volume.
2. shorten source file length.

What's the difference between Guennadi and Mine:

Mine: ipc3.c and ipc4.c is still the main source(separate driver interface), common is lib code.
Guennadi: ipc3.c and ipc4.c is lib ipc specific code, common is main source(ipc3/4 shared driver interface).

with this PR, common source have 300+ lines, and ipc3/ipc4 source have 600+ lines, compared with origin,  
it is much shorten and easy to read and maintain.

So I think this change is good, common is big or less actually is not the critical part of this PR.
Any future code or driver interface change will not impact another ipc, of course, you can change it as well if you want.

Reply:
> Hmm, looking at @lyakh 's code, so I understand right the proposal that we'd define function pointers for things like "vol_get_processing_function()" (with some optional parameters in the function template). So then we could have common code for e.g. prepare() (like in Guennadi's example), and only have the ops setup in ipc3/ipc4.c. I think this sounds reasonable, but devil may be in the details.

@kv2019i actually I meant it even simpler - just have two implementations of `vol_get_processing_function()` - one in volume_ipc3.c and one in volume_ipc4.c

Reply:
@kv2019i , @lyakh, no devils, it is not a difficult change, I just want to mix ipc3 and ipc4 code anymore.
If go with this way, any further ipc4 changes still need ipc3 verify, since they are still share prepare/params/proces, etc.